src/history/hs.c

```diff
@@ -392,7 +393,16 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
          */
         WT_TRET(__wt_cursor_key_order_init(cbt));
 #endif
-        WT_TRET(__hs_delete_key(session, cursor, btree->id, key));
+        session_flags = session->flags;
+        F_SET(session, WT_SESSION_IGNORE_HS_TOMBSTONE);
+        /* We're pointing at the newly inserted update. Iterate once more to avoid deleting it. */
+        ret = cursor->next(cursor);
+        if (ret == WT_NOTFOUND)
```

Comment:
The control flow doesn't look ideal to me here. I think this should become `WT_ERR_NOTFOUND_OK`. The use of `goto retry` makes that not very obvious, but I think we should add a level of indirection to resolve that, i.e: have `__hs_insert_record_with_btree` become everything up to the retry label plus:

while ((ret = __hs_insert_record_with_btree_internal()) == WT_RESTART) {}
return (ret);

Reply:
fb913dd